### PR TITLE
#6100 - create cacheable intermediate layer for S2I image

### PR DESCRIPTION
**Pull request by squakez**
#6100 The copy script would create a local cache with syndesis dependencies. This is filled in only the first time or every time a `mvn clean` is performed.

The docker image execute a `mvn verify` step to download dependencies (using cached syndesis jars, it would be the cached intermediate layer), and a `mvn package` that would take care to refresh any further dependency (or out of synch).

### CODE REVIEW

File: app/s2i/copy_mvn_repo.sh [hunk 1]
```diff
@@ -12,6 +12,19 @@ fi
 
 mkdir -p "${target_dir}/m2"
 cd ${local_repo}
+
+if [[ ! -d "${target_dir}/m2.cache" ]]; then
+    mkdir -p "${target_dir}/m2.cache"
```
Comment (nicolaferraro) [inline app/s2i/copy_mvn_repo.sh]: This should be idempotent, no need to check ;)

File: app/s2i/src/main/docker/Dockerfile [hunk 2]
```diff
@@ -34,4 +45,4 @@ RUN cd /tmp/artifacts/m2/project \
 RUN mkdir -p /opt/ipaas/
 COPY lic* /opt/ipaas/
 
-USER 1000
+USER 1000
```
Comment (nicolaferraro) [inline app/s2i/src/main/docker/Dockerfile]: Github likes newlines

File: app/s2i/copy_mvn_repo.sh [hunk 3]
```diff
@@ -12,6 +12,19 @@ fi
...
+fi
+
+if [[ ! "$(ls -A ${target_dir}/m2.cache)" ]]; then
+     echo "Going to cache locally Syndesis ${version} dependencies. Clean target directory to refresh the cache."
+     find ./io/syndesis | grep -F -- "/${version}/" | grep -v -- "-sources\." | grep -v -- "-tests\." | xargs tar -c | tar -vx -C "${target_dir}/m2.cache"
```
Comment (nicolaferraro) [inline app/s2i/copy_mvn_repo.sh]: Here if you initialize the `m2/.cache` with dependencies from your local maven (`mvn copy:dependencies`) you add the benefit of avoiding a "download the maven world", and will keep the benefits of layering.

**Comment by zregvart** [comment]
From what I see, this adds `target/m2.cache` and copies over Syndesis dependencies from `~/.m2/repository` and then within the Docker container runs a Maven build that uses this `m2.cache` directory copied over.
What I'm failing to understand is, how is this different from the build we already have, since it's doing much the same thing and the old bits are kept with this change?

**Comment by squakez** [comment]
> From what I see, this adds `target/m2.cache` and copies over Syndesis dependencies from `~/.m2/repository` and then within the Docker container runs a Maven build that uses this `m2.cache` directory copied over.
> What I'm failing to understand is, how is this different from the build we already have, since it's doing much the same thing and the old bits are kept with this change?

The problem reside in this statement https://github.com/syndesisio/syndesis/blob/master/app/s2i/src/main/docker/Dockerfile#L10
As the `copy_mvn_repo.sh` script always refresh snapshot dependencies, then Docker think it must recreate the below layers from scratch.

> From what I see, this adds `target/m2.cache` and copies over Syndesis dependencies from `~/.m2/repository` and then within the Docker container runs a Maven build that uses this `m2.cache` directory copied over.
> What I'm failing to understand is, how is this different from the build we already have, since it's doing much the same thing and the old bits are kept with this change?

The idea is that the `mvn verify` will be executed only the first time or at any target clean because we fix locally a copy of dependencies that are not going to be refreshed at every build.

**Comment by zregvart** [comment]
And as I understand I would need to run a clean build to get the S2I image updated?

If `-SNAPSHOT` dependency changed then I would expect that the docker image also needs to change. Perhaps we could refactor this to copy over the `-SNAPSHOT` dependencies to the last layer added instead?

I would suggest that we create the cached Maven repository outside of the docker build so it can take advantage of cached artefacts in the local Maven repository and copy the resulting repository to the image.

**Comment by nicolaferraro** [comment]
> And as I understand I would need to run a clean build to get the S2I image updated?
> 
> If `-SNAPSHOT` dependency changed then I would expect that the docker image also needs to change. Perhaps we could refactor this to copy over the `-SNAPSHOT` dependencies to the last layer added instead?
> 
> I would suggest that we create the cached Maven repository outside of the docker build so it can take advantage of cached artefacts in the local Maven repository and copy the resulting repository to the image.

I think that's almost what he's doing:
- phase 1: copy snapshots, download dependencies
- phase 2: override snapshots, download dependencies (if changed)

Phase 1 is skipped from second time you build.

As you said, if phase 1, instead of downloading dependencies, could copy them from local maven, this would be awesome, because it would improve build time, but it also allows to work with external (read Camel) snapshots.

**Comment by squakez** [comment]
> And as I understand I would need to run a clean build to get the S2I image updated?
> 

Not really, as the second build would bring the refreshed ones and would update accordingly. Let's say that the second build would take care of the incremental changes that there are between the 1.x-cached and the 1.x-refreshed. The first would bring majority of dependencies that you won't be touching. As in my case, working on a connector, I only refresh that connector dependencies, the rest won't be touched.

> If `-SNAPSHOT` dependency changed then I would expect that the docker image also needs to change. Perhaps we could refactor this to copy over the `-SNAPSHOT` dependencies to the last layer added instead?
> 
> I would suggest that we create the cached Maven repository outside of the docker build so it can take advantage of cached artefacts in the local Maven repository and copy the resulting repository to the image.

Yep, that's what @nicolaferraro was mentioning. I'm working right now on that part making sure that by running `mvn dependency:copy-dependencies` is still valid as a caching layer.

**Comment by squakez** [comment]
Just an head up after several experiments I've run. Actually there is a little limitation with the BOM dependencies, as they are not imported through the `mvn copy` plugin.

```
  <dependencyManagement>
    <dependencies>
      <dependency>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-dependencies</artifactId>
        <version>${spring-boot.version}</version>
        <type>pom</type>
        <scope>import</scope>
      </dependency>
      <dependency>
        <groupId>org.apache.camel</groupId>
        <artifactId>camel-spring-boot-dependencies</artifactId>
        <version>${camel.version}</version>
        <type>pom</type>
        <scope>import</scope>
      </dependency>
      <dependency>
        <groupId>io.syndesis.integration</groupId>
        <artifactId>integration-bom</artifactId>
        <version>${syndesis.version}</version>
        <type>pom</type>
        <scope>import</scope>
      </dependency>
    </dependencies>
  </dependencyManagement>
```
are not imported, so, unless copied through script (as we're doing for syndesis) the plugin won't work. I can issue a new request and work on that one separately from the issue I was solving here (docker caching layer).

**Comment by lburgazzoli** [comment]
@squakez what about using a custom maven settings like:

```xml
<?xml version="1.0" encoding="UTF-8" ?>
<settings 
  xmlns="http://maven.apache.org/SETTINGS/1.0.0" 
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" 
  xsi:schemaLocation="http://maven.apache.org/SETTINGS/1.0.0 http://maven.apache.org/xsd/settings-1.0.0.xsd">
    
  <localRepository>./target/repo</localRepository>
  
  <profiles>
    <profile>
      <id>_local</id>
      <activation>
        <activeByDefault>true</activeByDefault>
      </activation>
      <repositories>
        <repository>
          <id>_local</id>
          <url>file://${user.home}/.m2/repository</url>
          <releases>
            <enabled>true</enabled>
          </releases>
          <snapshots>
            <enabled>true</enabled>
          </snapshots>
        </repository>
      </repositories>
      <pluginRepositories>
        <pluginRepository>
          <id>_local</id>
          <url>file://${user.home}/.m2/repository</url>
          <releases>
            <enabled>true</enabled>
          </releases>
          <snapshots>
            <enabled>true</enabled>
          </snapshots>
        </pluginRepository>
      </pluginRepositories>
    </profile>
  </profiles>
</settings>
```

And:

```
mvn -s settings.xml  verify
```

**Comment by squakez** [comment]
I've created this issue to track the new requests: #6185 

The scope of this PR will remain according original #6100 - caching Docker layer.

**Comment by squakez** [comment]
Yeah, still have to commit the changes as suggested by @nicolaferraro. I am having some issue with minishift today, I'll push as soon I can confirm locally works fine. Thanks!

**Pull request merged**